src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    /// <summary>
+    /// In C#, tuples can be represented using tuple syntax and be given
+    /// names. However, the underlying representation for tuples unifies
+    /// to a single underlying tuple type, System.ValueTuple. Since the
+    /// names aren't part of the underlying tuple type they have to be
+    /// recorded somewhere else.
+    /// 
+    /// Roslyn records tuple names in an attribute: the
+    /// TupleElementNamesAttribute. The attribute contains a single string
+    /// array which records the names of the tuple elements in a breadth-first
+    /// traversal. If the type contains nested parameters, they are also recorded
```

Comment:
Yeah, Vlad's right: the difference is we just record "all" the names from the parent before traversing its element types. But that's just a pre-order traversal, since the element names belong to the parent tuple, not the elements.